datafusion/sql/src/utils.rs

```diff
@@ -255,3 +256,188 @@ pub(crate) fn normalize_ident(id: Ident) -> String {
+            // outer projection just select its name
+            let column_name = transformed_expr.display_name()?;
+            inner_projection_exprs.push(transformed_expr);
+            Ok(vec![Expr::Column(Column::from_name(column_name))])
+        }
+    } else {
+        Ok(vec![transformed_expr])
+    }
+}
+
+// write test for recursive_transform_unnest
+#[cfg(test)]
+mod tests {
+    use std::{ops::Add, sync::Arc};
+
+    use super::*;
+    use arrow::datatypes::{DataType as ArrowDataType, Field, Schema};
+    use arrow_schema::Fields;
+    use datafusion_common::DFSchema;
+    use datafusion_expr::EmptyRelation;
+
+    #[test]
+    fn test_recursive_transform_unnest() -> Result<()> {
+        let schema = Schema::new(vec![
+            Field::new(
+                "struct_col",
+                ArrowDataType::Struct(Fields::from(vec![
+                    Field::new("field1", ArrowDataType::Int32, false),
+                    Field::new("field2", ArrowDataType::Int32, false),
+                ])),
+                false,
+            ),
+            Field::new(
+                "array_col",
+                ArrowDataType::List(Arc::new(Field::new("item", DataType::Int64, true))),
+                true,
+            ),
+            Field::new("int_col", ArrowDataType::Int32, false),
+        ]);
+
+        let dfschema = DFSchema::try_from(schema)?;
+
+        let input = LogicalPlan::EmptyRelation(EmptyRelation {
+            produce_one_row: false,
+            schema: Arc::new(dfschema),
+        });
+
+        let mut unnest_placeholder_columns = vec![];
+        let mut inner_projection_exprs = vec![];
+
+        // unnest(struct_col)
+        let original_expr = Expr::Unnest(Unnest {
+            expr: Box::new(Expr::Column(Column::from_name("struct_col"))),
+        });
+        let transformed_exprs = recursive_transform_unnest(
+            &input,
+            &mut unnest_placeholder_columns,
+            &mut inner_projection_exprs,
+            original_expr,
+        )?;
+        assert_eq!(
+            transformed_exprs,
+            vec![
+                Expr::Column(Column::from_name("unnest(struct_col).field1")),
+                Expr::Column(Column::from_name("unnest(struct_col).field2")),
+            ]
+        );
+        assert_eq!(unnest_placeholder_columns, vec!["unnest(struct_col)"]);
+        // still reference struct_col in original schema but with alias,
+        // to avoid colliding with the projection on the column itself if any
+        assert_eq!(
+            inner_projection_exprs,
+            vec![
+                Expr::Column(Column::from_name("struct_col")).alias("unnest(struct_col)"),
+            ]
+        );
+
+        // unnest(array_col ) + 1
+        let original_expr = Expr::Unnest(Unnest {
+            expr: Box::new(Expr::Column(Column::from_name("array_col"))),
+        })
+        .add(Expr::Literal(ScalarValue::Int64(Some(1))));
```

Comment:
```suggestion
        .add(lit(1i64));
```

There are similar simplifications we could do below too